Sync up6 branch with master (#7040)
ballerina-platform/ballerina-dev-website: swan-lake/learn-the-platform/ballerina-persist/persist-cli.md

---
layout: ballerina-persist-documentation-left-nav-pages-swanlake
title: Persist CLI tool
description: The sections gives details on the Persist CLI tool.
keywords: ballerina, programming language, ballerina packages, persist, persist cli tool, persist init, persist generate, persist migrate
permalink: /learn/ballerina-persist/persist-cli/
active: persist_cli

> intro: One of the main components of the `bal persist` feature is the CLI tool. This tool is used to initialize the project with bal persist and generate the required files.

Author: ```suggestion
intro: One of the main components of the bal persist feature is the CLI tool. This tool is used to initialize the project with the bal persist command and generate the required files.
```
Backticks will not be rendered when added in the front matter. It will be fine to not have the backticks in the first sentence as it refers directly to the feature.  Shall we move the second sentence out of the intro as it refers to the command, which should be added within backticks for it to render properly?

redirect_from:
- /learn/ballerina-persist/persist-cli/
---

## Initialize bal persist in the project.

> The Ballerina project should be initialized with `bal persist` before generating the derived types, clients, and script files. This can be done using the `init` command. You can specify the preferred data store and the module which you need to generate files. If you do not specify the data store and the module, the default values will be used.

Author: ```suggestion
The Ballerina project should be initialized with `bal persist` before generating the derived types, clients, and script files. This can be done using the `init` command. You can specify the preferred data store and the module, which you need to generate files. If you do not specify the data store and the module, the default values will be used.
```

```bash
$ bal persist init --datastore="mysql" --module="store"
```

> | Command Parameter  |                                                     Description                                                      | Mandatory  |       Default Value        |

Author: ```suggestion
| Command parameter  |                                                     Description                                                      | Mandatory  |       Default value        |
```
We use sentence case in table headers also.

|:------------------:|:--------------------------------------------------------------------------------------------------------------------:|:----------:|:--------------------------:|
|     datastore      | used to indicate the preferred data store. Currently, three data stores are supported: inmemory, mysql, googlesheets |     No     |          immemory          |

> |       module       |                    used to indicate the persist enabled module in which the files are generated.                     |     No     |       <root_module>        |

Author: ```suggestion
|       `module`       |                    Used to indicate the persist-enabled module in which the files are generated.                     |     No     |       `<root_module>`        |
```

The command initializes `bal persist` feature in the project. This command will do the following,

1. Create persist directory in the project root directory.
   This directory contains the data model definition file(model.bal) of the project.
   
2. Create a model definition file in persist directory
   This file is used to model the data which need to be persisted in the project. You can have only one data model definition file in the project and file name can be anything with the `.bal` extension. The default file name is `model.bal`. Entities defined in this file should be based on the [`persist model` specification](persist_model.md).

3. Update the `Ballerina.toml` file with persist module configurations.

>    It will update the Ballerina.toml file with persist configurations, as follows,

Author: ```suggestion
   It will update the `Ballerina.toml` file with `persist` configurations as follows.
```

    ```ballerina
    [persist]
    datastore = "mysql"
    module = "rainier.store"
   ```

The directory structure will be,
```
rainier
   ├── persist
            └── model.bal
   ├── Ballerina.toml
   └── main.bal
```

Behaviour of the `init` command,

> - User should invoke the command within a Ballerina project

Author: ```suggestion
- You should invoke the command from within a Ballerina project.
```

- User can use the optional arguments to indicate the preferred module name and data store, otherwise default values will be used.

> - User cannot execute the command multiple times within the same project. User needs to remove the Ballerina.toml configurations, if the user wants to reinitialize the project.

Author: ```suggestion
- You cannot execute the command multiple times within the same project. You need to remove the `Ballerina.toml` configurations if the user wants to reinitialize the project.
```

## 3. Generating derived Types, client, and script files

> This command is used to generate the derived types, clients, and script files based on the data model definition file. This command is executed in the project root directory, as follows,

Author: ```suggestion
The command below is used to generate the derived types, clients, and script files based on the data model definition file. This command is executed from within the project root directory as follows.
```

```bash
$ bal persist generate
```

> It will add generated files under the `generated` directory. The directory structure will be as follows,

Author: ```suggestion
It will add the generated files under the `generated` directory. The directory structure will be as follows.
```

If the module name is provided, it will generate the files under a new subdirectory with the module name like below. Otherwise, it will generate the files under the `generated` directory.

```
rainier
   ├── generated
        └── store
              ├── persist_client.bal
              ├── persist_db_config.bal
              ├── persist_types.bal
              └── script.sql
   ├── persist
        └── model.bal
   ├── Ballerina.toml
   ├── Config.toml
   └── main.bal
```

> * The `persist_types.bal` file will contain the derived types based on the entities defined in the model definition file. 

Author: ```suggestion
* The `persist_types.bal` file will contain the derived types based on the entities defined in the model definition file. 
```
Shall we add these descriptions in a tabular format?

> * The `persist_client.bal` file will contain the client which is used to access the data store.

Author: ```suggestion
* The `persist_client.bal` file will contain the client, which is used to access the data store.
```

* The `persist_*_config.bal` file will contain the configurable variables required for the data store access. The configurable variables are changed based on the data store specified in the `Ballerina.toml` file. For example, if the data store is `mysql`, the configurable variables will be as follows.

```ballerina
configurable string host = ?;
configurable int port = ?;
configurable string user = ?;
configurable string password = ?;
configurable string database = ?;
```

> * The script file(script.sql) will contain the scripts to create the tables in the data store. This script file will be generated based on the data store specified in the `Ballerina.toml` file.

Author: ```suggestion
* The script file (`script.sql`) will contain the scripts to create the tables in the data store. This script file will be generated based on the data store specified in the `Ballerina.toml` file.
```

> Additionally, this command will create/update `Config.toml` file with configurables used to connect the client with the data store. Generated configurables will be based on the data store specified in the `Ballerina.toml` file.

Author: ```suggestion
Additionally, this command will create/update the `Config.toml` file with the configurables used to connect the client to the data store. The generated configurables will be based on the data store specified in the `Ballerina.toml` file.
```

> Behaviour of the `generate` command,

Author: ```suggestion
Behaviour of the `generate` command
```
Should this be a title?

> - User should invoke the command within a Ballerina project

Author: ```suggestion
- You should invoke the command from within a Ballerina project.
```

> - The user should have initiated `bal persist` feature in the project and update the model definition file.

Author: ```suggestion
- You should have the `bal persist` feature initiated in the project and update the model definition file.
```

- The model definition file should contain the `persist` module import (`import ballerina/persist as _;`)

> - The Model definition file should contain at least one entity

Author: ```suggestion
- The model definition file should contain at least one entity
```

> - If the user invokes the command twice, it will not fail. It will regenerate the files again.

Author: ```suggestion
- If you invoke the command twice, it will not fail. It will regenerate the files again.
```

> ## Generating Migration scripts for the Model Definition Changes[Experimental]

Author: ```suggestion
## Generate migration scripts for the model definition changes [Experimental]
```

> >**Info:** The support for migrations is currently an experimental feature, and its behavior may be subject to change in future releases. Also, the support for migrations is currently limited to MySQL data store.

Author: ```suggestion
>**Info:** The support for migrations is currently an experimental feature, and its behavior may be subject to change in future releases. Also, the support for migrations is currently limited to the MySQL data store.
```

> This command is used to generate the migration scripts for the model definition changes. This command is executed in the project root directory. This command will generate the migration scripts based on the changes in the model definition file. The generated migration scripts will be added to the `migrations` directory inside the `persist` directory.

Author: ```suggestion
The command below generates the migration scripts for the model definition changes. This command is executed in the project root directory. This command will generate the migration scripts based on the changes in the model definition file. The generated migration scripts will be added to the `migrations` directory inside the `persist` directory.
```

```bash
$ bal persist migrate add_employee
```

> The file structure of the project after executing the command will be as follows,

Author: ```suggestion
The file structure of the project after executing the command will be as follows.
```

```
rainier
   ├── generated
         └── store
               ├── persist_client.bal
               ├── persist_db_config.bal
               ├── persist_types.bal
               └── script.sql
   ├── persist
         └── model.bal
         └── migrations
               └── 20200820120000_add_employee
                  ├── model.bal
                  └── script.sql     
   ├── Ballerina.toml
   ├── Config.toml
   └── main.bal
```

> This command will create a new directory for the migration with the timestamp and the label provided by the user. The directory will contain the model definition file and the script file. The script file will contain the SQL script to update the database schema. The model definition file will be the snapshot of the model definition file at the time of executing the command. The next time the user executes the command, it will compare the model definition file with the latest snapshot file and generate the SQL script to update the database schema. So the user should not modify the snapshot file.

Author: ```suggestion
This command will create a new directory for the migration with the timestamp and the label provided by the user. The directory will contain the model definition file and the script file. The script file will contain the SQL script to update the database schema. The model definition file will be the snapshot of the model definition file at the time of executing the command. The next time the user executes the command, it will compare the model definition file with the latest snapshot file and generate the SQL script to update the database schema. So the user should not modify the snapshot file.
```
```suggestion
This command will create a new directory for the migration with the timestamp and the label provided by the user. The directory will contain the model definition file and the script file. The script file will contain the SQL script to update the database schema. The model definition file will be the snapshot of the model definition file at the time of executing the command. The next time you execute the command, it will compare the model definition file with the latest snapshot file and generate the SQL script to update the database schema. Therefore, you should not modify the snapshot file.
```

> Behaviour of the `migrate` command,

Author: ```suggestion
The behaviour of the `migrate` command will be as follows.
```
Normally, we say `In this code:`. For example, see http://localhost:3000/learn/work-with-data-using-queries-in-ballerina//

> - User should invoke the command within a Ballerina project

Author: ```suggestion
- You should invoke the command within a Ballerina project.
```

> - User should provide a valid label for the migration.

Author: ```suggestion
- You should provide a valid label for the migration.
```

> - The user should have initiated `bal persist` in the project and update the model definition file.

Author: ```suggestion
- You should have `bal persist` initiated in the project and the model definition file updated.
```

> - The user can execute the command multiple times. It will generate the migration scripts based on the changes in the model definition file.

Author: ```suggestion
- You can execute the command multiple times. It will generate the migration scripts based on the changes in the model definition file.
```
